README.md
# Show Hex dump String for binary file extension for Visual Studio Code

[//]: <> (## Features)

[//]: <> (\!\[feature X\]\(images/feature-x.png\))

[//]: <> (Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.)

## Usage

input `Select or Input` in the Command Palette, and select the `Select or Input absolute path of a binary file to show hex string`, to select or input the binary file path to show the hex dump string for the binary file

## Requirements

No pre-requirement.

## Extension Settings

This extension contributes the following settings:

* `show-hex-of-binary-file.pickType`: Set the binary file pick type, `select` to open file dialog to select on or multiple binary file, `input` to input the path
* `show-hex-of-binary-file.readOnly`: Set the binary file hex string show type, the default value is `true`
* `show-hex-of-binary-file.checkInterval`: Set check interval of the modification of the binary file in milliseconds, the minimal interval is 500, the default interval is 3000

[//]: <> (## Known Issues)

## Release Notes

### 0.0.1

Initial release of of this extension

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "show-hex-of-binary-file" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

out/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const vscode = require("vscode");
class HexStrProvider {
    constructor() {
        this._onDidChange = new vscode.EventEmitter();
    }
    get onDidChange() {
        return this._onDidChange.event;
    }
    update(uri) {
        this._onDidChange.fire(uri);
    }
    provideTextDocumentContent(uri, token) {
        let filePath = uri.fsPath;
        if (!filePath || filePath.trim() === '' || !fs.existsSync(filePath)) {
            return "";
        }
        let abspath = path.normalize(filePath);
        const baseNum = 16;
        let data = fs.readFileSync(filePath);
        let hexStr = '';
        for (let i = 0; i < data.length;) {
            if (i % baseNum === 0) {
                hexStr += i.toString(baseNum).padStart(8, '0') + ': ';
            }
            hexStr += data[i].toString(baseNum).padStart(2, '0');
            let j = 1;
            for (; j < baseNum; j++) {
                if ((i + j) >= data.length) {
                    break;
                }
                hexStr += ' ' + (data[i + j]).toString(baseNum).padStart(2, '0');
            }
            hexStr += '\n';
            i += baseNum;
        }
        return hexStr;
    }
}
exports.default = HexStrProvider;
HexStrProvider.scheme = 'hexstr';
//# sourceMappingURL=provider.js.map

out/extension.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const fs = require("fs");
const path = require("path");
const vscode = require("vscode");
const HexStrProvider_1 = require("./HexStrProvider");
const docList = new Map();
let hexStrProvider = new HexStrProvider_1.default();
let hexStrReadonly = true;
function getReadonlyCfg() {
    let cfgReadonly = vscode.workspace.getConfiguration().get("show-hex-of-binary-file.readOnly");
    if (typeof (cfgReadonly) === 'boolean') {
        hexStrReadonly = cfgReadonly;
    }
    return hexStrReadonly;
}
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "show-hex-of-binary-file" is now active!');
    // The command has been defined in the package.json file
    // Now provide the implementation of the command with registerCommand
    // The commandId parameter must match the command field in package.json
    let pickType = vscode.workspace.getConfiguration().get("show-hex-of-binary-file.pickType");
    getReadonlyCfg();
    pickType = pickType ? pickType : "select";
    let showBinHex;
    if (pickType === 'input') {
        showBinHex = inputToShowHex;
    }
    else {
        showBinHex = selectToShowHex;
    }
    let disposables = [
        vscode.workspace.registerTextDocumentContentProvider(HexStrProvider_1.default.scheme, hexStrProvider),
        vscode.commands.registerCommand('extension.showBinHexStr', showBinHex)
    ];
    context.subscriptions.push(...disposables);
    let doing = false;
    let closedDocs = [];
    let interval = vscode.workspace.getConfiguration().get("show-hex-of-binary-file.checkInterval");
    if (interval) {
        interval = interval < 500 ? 500 : interval;
    }
    else {
        interval = 3000;
    }
    setInterval(function () {
        if (doing) {
            return;
        }
        if (closedDocs.length > 0) {
            for (let i = 0; i < closedDocs.length; i++) {
                docList.delete(closedDocs[i]);
            }
            closedDocs = [];
        }
        doing = true;
        docList.forEach((docInfo, key) => __awaiter(this, void 0, void 0, function* () {
            // check each opened binary file every 1 second, and reload them if the source changed
            let doc = docInfo[0];
            if (doc.isClosed) {
                closedDocs.push(key);
            }
            else {
                let mtimeMs = docInfo[1];
                let size = docInfo[2];
                let filePath = key.substring(1);
                let newStat = fs.statSync(filePath);
                if (mtimeMs !== fs.statSync(filePath).mtimeMs || size !== newStat.size) {
                    console.log(`refresh the hex string of file: ${key} ${fs.statSync(filePath).mtimeMs} ${fs.statSync(filePath).size}`);
                    if (typeof (docInfo[3]) === 'boolean' && docInfo[3]) {
                        yield showHexStrPreview(filePath, doc);
                    }
                    else {
                        yield showHexStrFromBinFile(filePath, true);
                    }
                }
            }
        }));
        doing = false;
    }, interval);
}
exports.activate = activate;
function inputToShowHex() {
    return __awaiter(this, void 0, void 0, function* () {
        const filePath = yield vscode.window.showInputBox({
            placeHolder: 'Please input absolute path of the binary file to preview.',
        });
        if (filePath && filePath !== '') {
            if (fs.existsSync(filePath)) {
                let uri = vscode.Uri.file(filePath);
                try {
                    let doc = yield vscode.workspace.openTextDocument(uri);
                    vscode.window.showInformationMessage(`'${filePath}' isn't a binary file. It's a text file!`);
                }
                catch (e) {
                    // a binary file
                    if (e.toString().includes('File seems to be binary')) {
                        if (getReadonlyCfg()) {
                            showHexStrPreview(filePath);
                        }
                        else {
                            showHexStrFromBinFile(filePath);
                        }
                    }
                    else {
                        throw e;
                    }
                }
            }
            else {
                vscode.window.showInformationMessage(`Please input a valid absolute path of a binary file to preview!\n'${filePath}' isn't valid!`);
            }
        }
    });
}
/**
 * open file dialog to select binary file
 */
function selectToShowHex() {
    return __awaiter(this, void 0, void 0, function* () {
        const options = {
            canSelectMany: true,
            openLabel: 'Open',
            canSelectFiles: true,
            canSelectFolders: false
        };
        vscode.window.showOpenDialog(options).then((fileUris) => __awaiter(this, void 0, void 0, function* () {
            if (fileUris && fileUris[0]) {
                for (let i = 0; i < fileUris.length; i++) {
                    if (getReadonlyCfg()) {
                        yield showHexStrPreview(fileUris[i].fsPath);
                    }
                    else {
                        yield showHexStrFromBinFile(fileUris[i].fsPath);
                    }
                }
            }
        }));
    });
}
/**
 * show readonly preview editor for the binary file
 * @param filePath the absolute path of the binary file
 * @param alreadyOpen the already opened flag of the binary file
 */
function showHexStrPreview(filePath, openedDoc) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!filePath || filePath.trim() === '' || !fs.existsSync(filePath)) {
            return;
        }
        let abspath = path.normalize(filePath);
        let previewTitle = HexStrProvider_1.default.scheme + ':' + filePath;
        let uri = vscode.Uri.parse(previewTitle, false);
        if (openedDoc) {
            let stat = fs.statSync(abspath);
            docList.set('0' + abspath, [openedDoc, stat.mtimeMs, stat.size, true]);
            hexStrProvider.update(uri);
            return;
        }
        let opened = false;
        vscode.workspace.openTextDocument(uri).then(doc => {
            let stat = fs.statSync(abspath);
            docList.set('0' + abspath, [doc, stat.mtimeMs, stat.size, true]);
            vscode.window.showTextDocument(doc, { preview: false }).then(editor => {
                opened = true;
            });
        }, reason => {
            vscode.window.showInformationMessage(reason.toString());
        });
        while (!opened) {
            // wait the file open
            yield new Promise(done => setTimeout(done, 30));
        }
    });
}
/**
 * get the hex string of the binary file
 * @param filePath the file path of the binary file
 */
function getHexStrFromPath(filePath) {
    const baseNum = 16;
    let data = fs.readFileSync(filePath);
    let hexStr = '';
    for (let i = 0; i < data.length;) {
        if (i % baseNum === 0) {
            hexStr += i.toString(baseNum).padStart(8, '0') + ': ';
        }
        hexStr += data[i].toString(baseNum).padStart(2, '0');
        let j = 1;
        for (; j < baseNum; j++) {
            if ((i + j) >= data.length) {
                break;
            }
            hexStr += ' ' + (data[i + j]).toString(baseNum).padStart(2, '0');
        }
        hexStr += '\n';
        i += baseNum;
    }
    return hexStr;
}
/**
 * show hex string from the binary file
 * @param filePath the file path of the binary file
 * @returns void
 */
function showHexStrFromBinFile(filePath, alreadyOpen) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!filePath || filePath.trim() === '' || !fs.existsSync(filePath)) {
            return;
        }
        let abspath = path.normalize(filePath);
        let hexStr = getHexStrFromPath(filePath);
        let filename = path.basename(filePath);
        let previewTitle = 'untitled:' + filename;
        let uri = vscode.Uri.parse(previewTitle, false);
        let opened = false;
        vscode.workspace.openTextDocument(uri).then(doc => {
            let stat = fs.statSync(abspath);
            docList.set('1' + abspath, [doc, stat.mtimeMs, stat.size, false]);
            vscode.window.showTextDocument(doc).then(textEditor => {
                textEditor.edit(editBuilder => {
                    if (alreadyOpen) {
                        let range = new vscode.Range(new vscode.Position(0, 0), new vscode.Position(doc.lineCount, 57));
                        editBuilder.replace(range, hexStr);
                    }
                    else {
                        editBuilder.insert(new vscode.Position(0, 0), hexStr);
                    }
                    // flag that the file has been opened
                    opened = true;
                });
            });
        }, reason => {
            vscode.window.showInformationMessage(reason.toString());
        });
        while (!opened) {
            // wait the file open
            yield new Promise(done => setTimeout(done, 30));
        }
    });
}
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map

out/HexStrProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const vscode = require("vscode");
class HexStrProvider {
    constructor() {
        this._onDidChange = new vscode.EventEmitter();
    }
    get onDidChange() {
        return this._onDidChange.event;
    }
    update(uri) {
        this._onDidChange.fire(uri);
    }
    provideTextDocumentContent(uri, token) {
        let filePath = uri.fsPath;
        if (!filePath || filePath.trim() === '' || !fs.existsSync(filePath)) {
            return "";
        }
        const baseNum = 16;
        let data = fs.readFileSync(filePath);
        let hexStr = '';
        for (let i = 0; i < data.length;) {
            if (i % baseNum === 0) {
                hexStr += i.toString(baseNum).toUpperCase().padStart(8, '0') + ': ';
            }
            hexStr += data[i].toString(baseNum).toUpperCase().padStart(2, '0');
            let j = 1;
            for (; j < baseNum; j++) {
                if ((i + j) >= data.length) {
                    break;
                }
                hexStr += ' ' + (data[i + j]).toString(baseNum).toUpperCase().padStart(2, '0');
            }
            hexStr += '\n';
            i += baseNum;
        }
        return hexStr;
    }
}
exports.default = HexStrProvider;
HexStrProvider.scheme = 'hexstr';
//# sourceMappingURL=HexStrProvider.js.map

src/HexStrProvider.ts
import * as fs from 'fs';
import * as vscode from 'vscode';
export default class HexStrProvider implements vscode.TextDocumentContentProvider {
    static scheme = 'hexstr';
    constructor(public log: vscode.OutputChannel) {
        this._onDidChange = new vscode.EventEmitter<vscode.Uri>();
    }
    private _onDidChange: vscode.EventEmitter<vscode.Uri>;
    get onDidChange(): vscode.Event<vscode.Uri> {
        return this._onDidChange.event;
    }
    public update(uri: vscode.Uri) {
        this._onDidChange.fire(uri);
    }

    provideTextDocumentContent(uri: vscode.Uri, token: vscode.CancellationToken): string {
        let filePath = uri.fsPath;
        if (!filePath || filePath.trim() === '' || !fs.existsSync(filePath)) {
            return "";
        }
        let hexStr = HexStrProvider.getHexStrFromPath(filePath);
        this.log.appendLine(`Finish to retrive hex dump string of "${filePath}"`);
        return hexStr;
    }

    /**
     * get the hex string of the binary file
     * @param filePath the file path of the binary file
     */
    static getHexStrFromPath(filePath: string): string {
        const baseNum = 16;
        let data = fs.readFileSync(filePath);
        let hexStr = '';
        let byteCountPerLine = vscode.workspace.getConfiguration().get<number>("show-hex-of-binary-file.byteCountPerLine");
        if (typeof(byteCountPerLine) !== 'number' || byteCountPerLine%16 !== 0) {
            byteCountPerLine = 16;
        }
        for (let i=0; i<data.length;) {
            if (i%byteCountPerLine === 0) {
                hexStr += i.toString(baseNum).toUpperCase().padStart(8, '0') + ': ';
            }
            hexStr += data[i].toString(baseNum).toUpperCase().padStart(2, '0');
            let j = 1;
            for (;j<byteCountPerLine; j++) {
                if ((i+j) >= data.length) {
                    break;
                }
                hexStr += ' ' + (data[i+j]).toString(baseNum).toUpperCase().padStart(2, '0');
            }
            hexStr += '\n';
            i += byteCountPerLine;
        }
        return hexStr;
    }
}


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';
import * as child_process from 'child_process';
import HexStrProvider from './HexStrProvider';

const docList: Map<string, [vscode.TextDocument, number, number, boolean, vscode.TextEditor]> = new Map<string, [vscode.TextDocument, number, number, boolean, vscode.TextEditor]>();
let hexStrProvider: HexStrProvider;
let hexStrReadonly: boolean = true;
enum DocEditorState { Closed, UnChanged, Changed }
let log: vscode.OutputChannel;
let terminal: vscode.Terminal;

function getReadonlyCfg() {
	let cfgReadonly = vscode.workspace.getConfiguration().get<boolean>("show-hex-of-binary-file.readOnly");
	if (typeof(cfgReadonly) === 'boolean') {
		hexStrReadonly = cfgReadonly;
	}
	return hexStrReadonly;
}

/**
 * add menu event handler
 * @param uri the menu related doc uri
 * @param event the menu event
 */
function menuFunc(uri: vscode.Uri, event: vscode.Event<any>) {
	if (uri && uri.fsPath && fs.existsSync(uri.fsPath)) {
		if (getReadonlyCfg()) {
			showHexStrPreview(uri.fsPath);
		} else {
			showHexStrFromBinFile(uri.fsPath);
		}
	}
}
/**
 *
 * @param clickedUri the last clicked file
 * @param uris the all selected file
 */
function explorerMenuFunc(clickedUri: vscode.Uri, uris: vscode.Uri[]) {
	uris.forEach(async uri => {
		if (uri.scheme === 'file') {
			let filePath = uri.fsPath;
			let stat = fs.statSync(filePath);
			if (stat.isFile) {
				if (getReadonlyCfg()) {
					await showHexStrPreview(filePath);
				} else {
					await showHexStrFromBinFile(filePath);
				}
			}
		}
	});
}

function hasChanged(key: string, docInfo: [vscode.TextDocument, number, number, boolean, vscode.TextEditor], isTimer?: boolean): DocEditorState {
	if (!key || !docInfo || !docInfo[0] || !isTimer) {
		if (key) {
			docList.delete(key);
		}
		return DocEditorState.Closed;
	}
	let doc = docInfo[0];
	// log.appendLine(`${key.substring(1)} ${docInfo[0].isClosed}`);
	if (doc.isClosed) {
		docList.delete(key);
		return DocEditorState.Closed;
	}
	let mtimeMs = docInfo[1];
	let size = docInfo[2];
	let filePath = key.substring(1);
	let newStat:fs.Stats = fs.statSync(filePath);
	if (mtimeMs !== fs.statSync(filePath).mtimeMs || size !== newStat.size) {
		return DocEditorState.Changed;
	}
	return DocEditorState.UnChanged;
}

async function inputToExecuteInTerminal() {
	vscode.window.showInputBox({
		placeHolder: 'Please input the shell command you want to execute.',
	}).then(cmd => {
		if (cmd && cmd.trim() !== '') {
			if (!terminal) {
				terminal = vscode.window.createTerminal("jzhang test");
			}
			terminal.show();
			if (!terminal) {
				vscode.window.showErrorMessage("Failed to create Terminal!");
				return;
			}
			terminal.sendText(cmd);
		}
	});
}

async function inputToExecuteInNodejs() {
	vscode.window.showInputBox({
		placeHolder: 'Please input the shell command you want to execute and show result.',
	}).then(cmd => {
		if (cmd && cmd.trim() !== '') {
			child_process.exec(cmd, (error, stdout, stderr) => {
				let output: string = '';
				if (error) {
					output += error.toString() + "\n";
				}
				if (stdout) {
					output += stdout + "\n";
				}
				if (stderr) {
					output += stderr + "\n";
				}
				vscode.workspace.openTextDocument(vscode.Uri.parse('untitled:cmdResult')).then(doc=>{
					vscode.window.showTextDocument(doc).then(textEditor => {
						textEditor.edit(editBuilder => {
							editBuilder.insert(new vscode.Position(doc.lineCount, 0), output);
						});
					});
				});
			});
		}
	});
}

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	log = vscode.window.createOutputChannel("HexDumStr");
	hexStrProvider = new HexStrProvider(log);
	log.appendLine('Congratulations, your extension "show-hex-of-binary-file" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json

	let pickType = vscode.workspace.getConfiguration().get<string>("show-hex-of-binary-file.pickType");
	getReadonlyCfg();
	pickType = pickType ? pickType : "select";
	let showBinHex;
	if (pickType === 'input') {
		showBinHex = inputToShowHex;
	} else {
		showBinHex = selectToShowHex;
	}
	let disposables: vscode.Disposable[] = [
		vscode.workspace.registerTextDocumentContentProvider(HexStrProvider.scheme, hexStrProvider),
		vscode.commands.registerCommand('extension.showBinHexStr', showBinHex),
		vscode.commands.registerCommand('editor.title.showBinHexStr', menuFunc),
		vscode.commands.registerCommand('editor.context.showBinHexStr', menuFunc),
		vscode.commands.registerCommand('editor.title.context.showBinHexStr', menuFunc),
		vscode.commands.registerCommand('explorer.context.showBinHexStr', explorerMenuFunc),
		vscode.commands.registerCommand('extension.execute.interminal', inputToExecuteInTerminal),
		vscode.commands.registerCommand('extension.execute.innodejs', inputToExecuteInNodejs),
	];

	context.subscriptions.push(...disposables);

	let doing = false;
	let closedDocs: string[] = [];
	let interval = vscode.workspace.getConfiguration().get<number>("show-hex-of-binary-file.checkInterval");
	if (interval) {
		interval = interval < 500 ? 500: interval;
	} else {
		interval = 3000;
	}
	setInterval(function(){
		if (doing) {
			return;
		}
		if (closedDocs.length > 0) {
			for (let i=0; i<closedDocs.length; i++) {
				docList.delete(closedDocs[i]);
			}
			closedDocs = [];
		}
		doing = true;
		docList.forEach(async (docInfo, key) => {
			// check each opened binary file every 1 second, and reload them if the source changed
			let stat = hasChanged(key, docInfo, true);
			if (stat === DocEditorState.Changed) {
				let filePath = key.substring(1);
				log.appendLine(`refresh the hex string of file: ${key} ${fs.statSync(filePath).mtimeMs} ${fs.statSync(filePath).size}`);
				if (typeof(docInfo[3]) === 'boolean' && docInfo[3]) {
					await showHexStrPreview(filePath, docInfo[0]);
				} else {
					await showHexStrFromBinFile(filePath, true);
				}
			} else if (stat === DocEditorState.Closed) {
				closedDocs.push(key);
			}
		});
		doing = false;
	}, interval);
}

/**
 * show input box to input the absolute path of the binary file.
 */
async function inputToShowHex() {
	const filePath = await vscode.window.showInputBox({
		placeHolder: 'Please input absolute path of the binary file to preview.',
	});
	if (filePath) {
		showHexStrFromBinFile(filePath);
	}
}
/**
 * open file dialog to select binary file
 */
async function selectToShowHex() {
	const options: vscode.OpenDialogOptions = {
		canSelectMany: true,
		openLabel: 'Open',
		canSelectFiles: true,
		canSelectFolders: false
   	};
   	vscode.window.showOpenDialog(options).then(async fileUris => {
	   	if (fileUris && fileUris[0]) {
			for (let i=0; i<fileUris.length; i++) {
				if (getReadonlyCfg()) {
					await showHexStrPreview(fileUris[i].fsPath);
				} else {
					await showHexStrFromBinFile(fileUris[i].fsPath);
				}
			}
	   	}
   	});
}
/**
 * show readonly preview editor for the binary file
 * @param filePath the absolute path of the binary file
 * @param alreadyOpen the already opened flag of the binary file
 */
async function showHexStrPreview(filePath: string, openedDoc?: vscode.TextDocument): Promise<void> {
	if (!filePath || filePath.trim() === '' || !fs.existsSync(filePath)) {
		return;
	}
	let abspath = path.normalize(filePath);
	let key = '0' + abspath;
	if (!openedDoc) {
		// get the opened doc info list to check whether it has already been opended
		if (docList.has(key)) {
			let docInfo = docList.get(key);
			if (docInfo && docInfo[0]) {
				if (docInfo[0].isClosed) {
					docList.delete(key);
				} else {
					let stat = hasChanged(key, docInfo);
					if (stat === DocEditorState.Changed) {
						openedDoc = docInfo[0];
					} else if (stat === DocEditorState.UnChanged) {
						// log.appendLine(`The content hasn't been modified! "${abspath}"`);
						return;
					}
				}
			}
		}
	}
	let previewTitle: string = HexStrProvider.scheme + ':' + filePath.replace(/\\/g, '/');
	let uri: vscode.Uri = vscode.Uri.parse(previewTitle, false);
	if (openedDoc) {
		let stat:fs.Stats = fs.statSync(abspath);
		let docInfo = docList.get(key);
		if (docInfo) {
			docList.set(key, [openedDoc, stat.mtimeMs, stat.size, true, docInfo[4]]);
		}
		hexStrProvider.update(uri);
		return;
	}
	let opened = false;
	let stat:fs.Stats = fs.statSync(abspath);
	let loopInterval = 30;
	if (stat.size > 200000) {
		// let loop interval as 1 second if the file size is larger than 200 kb
		loopInterval = 1000;
	}
	vscode.workspace.openTextDocument(uri).then(doc => {
		vscode.window.showTextDocument(doc, {preview: false}).then(editor => {
			docList.set(key, [doc, stat.mtimeMs, stat.size, true, editor]);
			opened = true;
		}, err => {
			log.appendLine(err);
			opened = true;
		});

	}, reason => {
		vscode.window.showInformationMessage(reason.toString());
	});
	while (!opened) {
		// wait the file open
		await new Promise(done => setTimeout(done, loopInterval));
	}
}

/**
 * show hex string from the binary file
 * @param filePath the file path of the binary file
 * @returns void
 */
async function showHexStrFromBinFile(filePath: string, alreadyOpen?: boolean): Promise<void> {
	if (!filePath || filePath.trim() === '' || !fs.existsSync(filePath)) {
		return;
	}
	let abspath = path.normalize(filePath);
	let key = '1' + abspath;
	if (!alreadyOpen) {
		// get the opened doc info list to check whether it has already been opended
		if (docList.has(key)) {
			let docInfo = docList.get(key);
			if (docInfo && docInfo[0]) {
				if (docInfo[0].isClosed) {
					docList.delete(key);
				} else {
					let stat = hasChanged(key, docInfo);
					if (stat === DocEditorState.Changed) {
						alreadyOpen = true;
					} else if (stat === DocEditorState.UnChanged) {
						// log.appendLine(`The content hasn't been modified! "${abspath}"`);
						return;
					}
				}
			}
		}
	}
	let hexStr = HexStrProvider.getHexStrFromPath(filePath);
	let filename = path.basename(filePath);
	let previewTitle: string = 'untitled:' + filename;
	let uri: vscode.Uri = vscode.Uri.parse(previewTitle, false);
	let opened = false;
	vscode.workspace.openTextDocument(uri).then(doc => {
		let stat:fs.Stats = fs.statSync(abspath);
		vscode.window.showTextDocument(doc).then(textEditor => {
			docList.set(key, [doc, stat.mtimeMs, stat.size, false, textEditor]);
			textEditor.edit(editBuilder =>{
				if (alreadyOpen) {
					let range = new vscode.Range(new vscode.Position(0, 0), new vscode.Position(doc.lineCount, 57));
					editBuilder.replace(range, hexStr);
				} else {
					editBuilder.insert(new vscode.Position(0, 0), hexStr);
				}
				// flag that the file has been opened
				opened = true;
			});
		});
	}, reason => {
		vscode.window.showInformationMessage(reason.toString());
	});
	while (!opened) {
		// wait the file open
		await new Promise(done => setTimeout(done, 30));
	}
}

// this method is called when your extension is deactivated
export function deactivate() {
	if (terminal) {
		terminal.dispose();
	}
}
